appserver/payara-appserver-modules/microprofile/fault-tolerance/src/main/java/fish/payara/microprofile/faulttolerance/policy/FaultTolerancePolicy.java

```diff
@@ -0,0 +1,609 @@
+            }
+            logger.log(Level.FINE, "Using fallback method: {0}", fallback.method.getName());
+            return invocation.service.fallbackInvoke(fallback.method, invocation.context);
+        } finally {
+            invocation.endTrace();
+        }
+    }
+
+    /**
+     * Stage that takes care of the {@link RetryPolicy} handling.
+     */
+    private Object processRetryStage(FaultToleranceInvocation invocation) throws Exception {
+        if (!retry.isNone()) {
+            logger.log(Level.FINER, "Proceeding invocation with retry semantics");
+        }
+        int totalAttempts = retry.totalAttempts();
+        int attemptsLeft = totalAttempts;
+        Long retryTimeoutTime = retry.timeoutTimeNow();
+        while (attemptsLeft > 0) {
+            attemptsLeft--;
+            try {
+                boolean firstAttempt = attemptsLeft == totalAttempts - 1;
+                if (!firstAttempt) {
+                    logger.log(Level.FINER, "Attempting retry.");
+                    invocation.metrics.incrementRetryRetriesTotal();
+                }
+                Object resultValue = isAsynchronous() 
+                        ? processRetryAsync(invocation)
+                        : processCircuitBreakerStage(invocation, null);
+                if (firstAttempt) {
+                    invocation.metrics.incrementRetryCallsSucceededNotRetriedTotal();
+                } else {
+                    invocation.metrics.incrementRetryCallsSucceededRetriedTotal();
+                }
+                return resultValue;
+            } catch (Exception ex) {
+                boolean timedOut = retryTimeoutTime != null && System.currentTimeMillis() >= retryTimeoutTime;
+                if (attemptsLeft <= 0 || !retry.retryOn(ex) || timedOut) {
+                    logger.log(Level.FINE, "Retry attemp failed. Giving up{0}", timedOut ? " due to time-out." : ".");
+                    invocation.metrics.incrementRetryCallsFailedTotal();
+                    throw ex;
+                }
+                logger.log(Level.FINE, "Retry attempt failed. {0} attempts left.", attemptsLeft);
+                if (retry.isDelayed()) {
+                    invocation.service.delay(retry.jitteredDelay(), invocation.context);
+                }
+            }
+        }
+        // this line should never be reached as we throw above
+        throw new FaultToleranceException("Retry failed"); 
+    }
+
+    private CompletableFuture<Object> processRetryAsync(FaultToleranceInvocation invocation) throws Exception {
+        CompletableFuture<Object> asyncAttempt = new CompletableFuture<>();
+        invocation.service.runAsynchronous(asyncAttempt, invocation.context,
+                () -> invocation.runStageWithWorker(() -> processCircuitBreakerStage(invocation, asyncAttempt)));
+        try {
+            asyncAttempt.get(); // wait and only proceed on success
```

Comment:
The timeout is in the timeout stage. It will complete the `CompletableFuture` and thereby effectively make this `get` into one that will timeout if timeout policy is used. This is important as this is the only way to unravel the stages correctly in case of a timeout since there is the circuit breaker stage between the retry and the timeout.